### PR TITLE
add github auth login

### PR DESCRIPTION
**Pull request by JiahuiZhao11**
Ⅰ. Describe what this PR does
add github auth login
Ⅱ. Does this pull request fix one issue?
no
Ⅲ. List the added test cases (unit test/integration test) if any, please explain if no tests are needed.

Ⅳ. Describe how to verify it

Ⅴ. Special notes for reviews

### CODE REVIEW

File: deploy/template/authConfigmap.yaml [hunk 1]
```diff
@@ -0,0 +1,9 @@
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  name: kubecube-auth-config
+  namespace: kubecube-system
+data:
+  github.enabled: enabled
+  github.clientId:
```
Comment (weilaaa) [inline deploy/template/authConfigmap.yaml:8]: What if add another way to oAuth? How do we organize the structure to make this look clearer？

File: pkg/apiserver/cubeapi/user/auth.go [hunk 2]
```diff
@@ -156,3 +119,123 @@ func Login(c *gin.Context) {
...
+func GitHubLogin(c *gin.Context) {
+	code := c.Query("code")
+	if code == "" {
+		clog.Error("code is null")
+		response.FailReturn(c, errcode.AuthenticateError)
+		return
+	}
+
+	g := github.GetProvider()
+	if !g.GitHubIsEnable {
```
Comment (weilaaa) [inline pkg/apiserver/cubeapi/user/auth.go:132]: What is g? Is that can be nil?

File: pkg/apiserver/cubeapi/user/auth.go [hunk 3]
```diff
@@ -156,3 +119,123 @@ func Login(c *gin.Context) {
...
+
+	userInfo, err := g.IdentityExchange(code)
+	if err != nil {
+		response.FailReturn(c, errcode.AuthenticateError)
+		return
+	}
+	clog.Info("user %s auth success by github", userInfo.GetUserName())
+
+	// get user by name
+	userName := hex.EncodeToString([]byte(userInfo.GetUserName()))
```
Comment (weilaaa) [inline pkg/apiserver/cubeapi/user/auth.go:146]: Why we need hex?

File: pkg/apiserver/cubeapi/user/auth.go [hunk 4]
```diff
@@ -156,3 +119,123 @@ func Login(c *gin.Context) {
...
+		response.FailReturn(c, errcode.UserIsDisabled)
+		return
+	}
+
+	// if user first login, create user
+	user := &v1.User{}
+	if userFind == nil {
+		user.Name = gitHubUserNamePrefix + userName
+		user.Spec.LoginType = v1.GitHubLogin
+		user.Spec.Password = md5util.GetMD5Salt(uuid.New().String())
```
Comment (weilaaa) [inline pkg/apiserver/cubeapi/user/auth.go:162]: We are supposed to add rand seed if we want to use random

Comment (weilaaa) [inline pkg/apiserver/cubeapi/user/auth.go:162]: We are supposed to add rand seed if we want to use random(rand.Seed(time.Now().UnixNano()))

File: pkg/apiserver/cubeapi/user/auth.go [hunk 5]
```diff
@@ -156,3 +119,123 @@ func Login(c *gin.Context) {
...
+	respInfo = UpdateUserStatusImpl(c, user)
+	if respInfo != nil {
+		response.FailReturn(c, respInfo)
+		return
+	}
+
+	// generate token and return
+	token, respInfo := jwt.GenerateToken(userName, 0)
+	bearerToken := jwt.BearerTokenPrefix + " " + token
+	if respInfo != nil {
```
Comment (weilaaa) [inline pkg/apiserver/cubeapi/user/auth.go:183]: why don't we use err instead of respInfo?

File: pkg/apiserver/cubeapi/user/auth.go [hunk 6]
```diff
@@ -156,3 +119,123 @@ func Login(c *gin.Context) {
...
+	// ldap login
+	ldapProvider := ldap.GetProvider()
+	userInfo, err := ldapProvider.Authenticate(name, password)
+	if err != nil {
+		return nil, errcode.AuthenticateError
+	}
+	clog.Info("user %s auth success by ldap", userInfo.GetUserName())
+
+	// if user first login, create user
+	if user == nil {
```
Comment (weilaaa) [inline pkg/apiserver/cubeapi/user/auth.go:232]: wtf

File: pkg/authentication/config.go [hunk 7]
```diff
@@ -49,3 +50,9 @@ type GenericConfig struct {
 	Method              string `yaml:"method,omitempty"`
 	GenericAuthIsEnable bool   `yaml:"genericAuthIsEnable, omitempty"`
 }
+
+type GitHubConfig struct {
+	GitHubIsEnable bool
+	ClientID       string `yaml:"clientID,omitempty"`
```
Comment (weilaaa) [inline pkg/authentication/config.go:56]: Why don't use json tag?

File: pkg/authentication/identityprovider/github/config.go [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+	v1 "k8s.io/api/core/v1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+const configMapName = "kubecube-auth-config"
+
+func getConfig() authentication.GitHubConfig {
+	var config authentication.GitHubConfig
+
+	kClient := clients.Interface().Kubernetes(constants.PivotCluster).Cache()
```
Comment (weilaaa) [inline pkg/authentication/identityprovider/github/config.go:20]: We should determine if client of pivot cluster is nil

File: pkg/authentication/identityprovider/github/config.go [hunk 9]
```diff
@@ -0,0 +1,36 @@
...
+	cm := &v1.ConfigMap{}
+	err := kClient.Get(context.Background(), client.ObjectKey{Name: configMapName, Namespace: constants.CubeNamespace}, cm)
+	if err != nil {
+		clog.Error("get configmap from K8s err: %v", err)
+		return config
+	}
+
+	if cm.Data["github.enabled"] == "true" {
```
Comment (weilaaa) [inline pkg/authentication/identityprovider/github/config.go:28]: What would happened if Data has no key that "github.enabled"

File: pkg/authentication/identityprovider/github/config.go [hunk 10]
```diff
@@ -0,0 +1,36 @@
...
+		config.GitHubIsEnable = true
+	} else {
+		config.GitHubIsEnable = false
+	}
+	config.ClientID = cm.Data["github.clientId"]
+	config.ClientSecret = cm.Data["github.clientSecret"]
```
Comment (weilaaa) [inline pkg/authentication/identityprovider/github/config.go:34]: same as above

File: pkg/authentication/identityprovider/github/github.go [hunk 11]
```diff
@@ -99,35 +90,81 @@ func (g githubIdentity) GetUserEmail() string {
...
+
+func (g *githubProvider) IdentityExchange(code string) (identityprovider.Identity, error) {
+	if g.ClientID == "" || g.ClientSecret == "" {
+		clog.Error("clientId or clientSecret is null")
+		return nil, errors.New("clientId or clientSecret is null")
 	}
-	token, err := g.Config.Exchange(ctx, code)
+
+	// get token
+	req, err := http.NewRequest(http.MethodPost, "https://github.com/login/oauth/access_token?client_id="+g.ClientID+"&client_secret="+g.ClientSecret+"&code="+code, nil)
```
Comment (weilaaa) [inline pkg/authentication/identityprovider/github/github.go:109]: Use fmt.Sprintf() instead of it

**Review by weilaaa — COMMENTED** [review]
need fixed

File: pkg/apiserver/cubeapi/user/auth.go (hunk 3 above)
Comment (JiahuiZhao11) [reply pkg/apiserver/cubeapi/user/auth.go:146]: github name may not be allowed to store to K8s，such as capital letters

**Review by fusida — APPROVED** [review]
LGTM

**Pull request merged**